plugins/search/src/components/SearchPage/SearchPage.tsx

```diff
@@ -13,55 +13,57 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-import { Content, Header, Page, useQueryParamState } from '@backstage/core';
-import { Grid } from '@material-ui/core';
-import React, { useEffect, useState } from 'react';
-import { useDebounce } from 'react-use';
-import { SearchBar } from '../SearchBar';
-import { SearchResult } from '../SearchResult';
 
-export const SearchPage = () => {
-  const [queryString, setQueryString] = useQueryParamState<string>('query');
-  const [searchQuery, setSearchQuery] = useState(queryString ?? '');
-
-  const handleSearch = (event: React.ChangeEvent<HTMLInputElement>) => {
-    event.preventDefault();
-    setSearchQuery(event.target.value);
-  };
+import React from 'react';
+import qs from 'qs';
+import { useLocation, useOutlet } from 'react-router';
+import { SearchContextProvider, useSearch } from '../SearchContext';
+import { JsonObject } from '@backstage/config';
+import { LegacySearchPage } from '../LegacySearchPage';
 
-  useEffect(() => setSearchQuery(queryString ?? ''), [queryString]);
+export const UrlUpdater = () => {
+  const { term, types, pageCursor, filters } = useSearch();
 
-  useDebounce(
-    () => {
-      setQueryString(searchQuery);
+  const newParams = qs.stringify(
+    {
+      query: term,
+      types,
+      pageCursor,
+      filters,
     },
-    200,
-    [searchQuery],
+    { arrayFormat: 'brackets' },
   );
+  const newUrl = `${window.location.pathname}?${newParams}`;
+
+  // We directly manipulate window history here in order to not re-render
+  // infinitely (state => location => state => etc). The intention of this
```

Comment:
that being said, the `location => state` direction should probably strictly only happen on first mount to feed initial state values, otherwise I think the flow of events gets very confusing (plus can cause issues as mentioned). So doing "hard" setting like this can sometime have its place, but should not be done for the reason of breaking refresh loops but to handle the history stack in a certain way (e.g. avoiding a history event for every keypress)